P2461.md

```diff
@@ -605,37 +735,67 @@ We therefore have a plausible RAII-based metaphor that people already understand
 
-# What about abbreviated lambdas?
+## Capture design space
 
-The `post`-condition syntax naturally looks like a shorthand lambda:
+There is design space in what kind of captures we allow of parameters of different kinds.
 
-```cpp
-post [ @_closure_@ ] ( @_identifier_@ ) { @_conditional-expression_@ }
-```
+The case space spans between condition (`pre`, `post`) by parameter kind
+(value, lvalue, rvalue) by qualifier (non-const, `const`) by kind-of-capture
+(reference, value), which gives us $2*3*2*2 = 24$ cases.
 
-This topic was explored in [@P0573R2] by Barry Revzin, who proposed this syntax as point 2.3.
-
-Alternative from the paper:
+The example is
 
 ```cpp
-post [ @_closure_@ ] ( @_identifier_@ ) => @_conditional-expression_@
+auto f(@_param_@ x)
+    pre @_cap_@ { x }
+    post @_cap_@ { x }
+;
 ```
 
-However, this doesn't work in function declarations because terminating
-expressions is difficult, so we like the earlier syntax better.
-
-In this case, the default closure is `[&]`, the parameter-type is `auto&&` and
-we don't need to take a stance on the `noexcept` specifier or the return type.
-
-However, given the above, EWG should take a stance on whether that looks like a
-plausible set of semantics for shorthand lambdas, because we shouldn't have a
-yet another set of semantics for those.
+| #  | param        | cap   | cond   | MVP | Ext  |
+| -- | ------------ | ----- | ------ | --- | ---- |
+|  1 | `T`          | `[&]` | `pre`  | y   | y    | 
+|  2 | `T`          | `[&]` | `post` | (!) | (!)  |
+|  3 | `T`          | `[=]` | `pre`  |     | y    |
+|  4 | `T`          | `[=]` | `post` |     | y    |
+|  5 | `T const`    | `[&]` | `pre`  | y   | y    |
+|  6 | `T const`    | `[&]` | `post` | y   | y    |
+|  7 | `T const`    | `[=]` | `pre`  |     | y    |
+|  8 | `T const`    | `[=]` | `post` |     | y    |
+|  9 | `T &`        | `[&]` | `pre`  | y   | y    |
+| 10 | `T &`        | `[&]` | `post` | y   | y    |
+| 11 | `T &`        | `[=]` | `pre`  |     | y    |
+| 12 | `T &`        | `[=]` | `post` |     | y    |
+| 13 | `T const &`  | `[&]` | `pre`  | y   | y    |
+| 14 | `T const &`  | `[&]` | `post` | y   | y    |
+| 15 | `T const &`  | `[=]` | `pre`  |     | y    |
+| 16 | `T const &`  | `[=]` | `post` |     | y    |
+| 17 | `T &&`       | `[&]` | `pre`  | y   | y    |
+| 18 | `T &&`       | `[&]` | `post` | y   | y    |
+| 19 | `T &&`       | `[=]` | `pre`  |     | y    |
+| 20 | `T &&`       | `[=]` | `post` |     | y    |
+| 21 | `T const &&` | `[&]` | `pre`  | y   | y    |
+| 22 | `T const &&` | `[&]` | `post` | y   | y    |
+| 23 | `T const &&` | `[=]` | `pre`  |     | y    |
+| 24 | `T const &&` | `[=]` | `post` |     | y    |
+
+- **MVP**: this paper.
+- **Ext**: this paper plus minimal "allow captures" extension.
+- **(!)**: Capturing by value in postconditions makes no sense to the caller,
+  therefore it's not a well-behaved contract and we should diagnose (2).
```

Comment:
Am I misreading something? Isn't (2) capture-by-reference?

I think you mean to say here that capturing a non-const by reference in postconditions makes no sense to the caller. Capturing by value I think is the solution, no?

Reply:
Fixed.